### ISSUE TITLE
composability issue with class based hocs

### ISSUE THREAD

**Issue by ioi0** [zanettin/incompose#26]
This issue aligns with https://github.com/zanettin/incompose/issues/23

I've been using the library in production for a pretty decent amount of time (big thanks for this 👍🏻). But I've ran into composability issues occasionally. 

Such issues occur when consumer tries to compose multiple hocs one of which has "class" as a return value while other ones attempt to use it as a function. 

For clarity here are some hocs that return "classes": `withReducer`, `withPropsOnChange`, `withLifecycle`

If you'll try to compose those with other hocs such as `withProps` or `withHandlers` you'll run into issue.

An example bellow will fail as https://github.com/zanettin/incompose/issues/23

```
const Composed = compose(
  withProps({ foo: 'bar'  }),
  withPropsOnChange(() => true, props => props),
)(() => <span>Template</span>);
``` 

Next one will fail `Uncaught TypeError: Cannot set property 'state' of undefined`

```
const Composed = compose(
  withProps({ foo: 'bar'  }),
  withLifecycle({ componentDidMount: () => console.log('Hello') }),
)(() => <span>Template</span>);
``` 

I've expected those functions to works nicely together. 
Please, correct me if I am wrong or missing the point behind this library design.

I think there are two way we can approach this issue. 

1. Either by updating first set of hocs in the same manner as in https://github.com/zanettin/incompose/issues/23

With this solution we kinda set a convention that every input component in composition _should be_ functional. The drawback - we have to care about it manually

2. update another set of hocs replacing component function expressions into jsx (`Component({ ...config, ...props })` into `<Component {...config} {...props} />`) or by using `createElement` Inferno API

With this solution we delegate the issues to Inferno layer

Let me know which solution you think is more preferable?

**Comment by zanettin** [comment]
Hi @ioi0 

Thanks again for the kind words and your inputs.
haven't worked with inferno for a while so i've missed some updates on the core 🙈  regarding your questions: yes, they should (still) be composable and yes, the JSX pattern is the one i'd go with.

i've started a new PR (#27 ) over there and played around with almost all available HoCs and it does work so far. my test code can be found below:

```
import {
  compose,
  withProps,
  withPropsOnChange,
  withLifecycle,
  withHandlers,
  branch,
  renderComponent,
  defaultProps,
  flattenProps,
  mapProps,
  renameProp,
  renameProps,
  withState,
  shouldUpdate
} from "incompose";
import { linkEvent } from "incompose/node_modules/inferno";

const App = () => null;

const inc = props => props.setCounter(props.counter + 1);

const Counter = props => (
  <div>
    <p>counter: {props.counter}</p>
    <button onClick={linkEvent(props, inc)}>+</button>
  </div>
);

export default compose(
  withState("counter", "setCounter", 1),
  withProps({
    foo: "bar",
    person: { firstName: "John", lastName: "Doe", age: 33 },
    price: 200,
    discount: 20
  }),
  defaultProps({ foo: 123, bar: 321 }), // should not override withProps
  flattenProps("person"),
  mapProps(({ price, discount, ...props }) => ({
    discountPrice: price - discount,
    ...props
  })),
  renameProp("firstName", "first-name"),
  renameProps({ lastName: "last-name", age: "a-g-e" }),
  withPropsOnChange(() => true, props => props),
  withLifecycle({
    componentDidMount: () => console.log("did mount"),
    componentWillMount: () => console.log("will mount"),
    componentDidUpdate: () => console.log("did update"),
    componentWillUpdate: () => console.log("will update"),
    componentWillUnmount: () => console.log("will unmount")
  }),
  withHandlers({
    getRandomValue: _ => Math.floor(Math.random() * 1000)
  }),
  shouldUpdate((props, nextProps) => {
    console.warn("props", props);
    console.warn("nextProps", nextProps);
    return true;
  }),
  branch(
    ({ getRandomValue }) => getRandomValue % 2 === 0,
    renderComponent(props => {
      console.warn("left branch", props);
      return <Counter {...props} />;
    }),
    renderComponent(props => {
      console.warn("right branch", props);
      return <Counter {...props} />;
    })
  )
)(App);
```

feel free to check out this PR and use it with  `yarn link` and `yarn link incompose` on your project. would be glad to get some feedback from you.

NOTE: i've also switched the implementation of `withLifecycle` and `shouldUpdate` to use the hook api from inferno. not sure since when these are available but i've tested my code with inferno 7.1.13.

**Comment by zanettin** [comment]
v5.0.1 shipped 🎉 
https://www.npmjs.com/package/incompose
thanks again for your help @ioi0 👍
